### PR TITLE
[GOBBLIN-1630] Remove flow level metrics for adhoc flows

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1630

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Gobblin-as-a-Service emits flow level events for users to keep track of their flow compilation status, job status, number of workunits etc. However, users rarely if ever track adhoc-flows, where they are generally used in a fire-and-forget manner. Furthermore, keeping track of metrics continuously adds strains to the metric libraries which leads to wasted resources.

We should only be emitting flow level events for flows that are scheduled and saved.

This PR adds a flag for adhoc flows such that downstream executions can also differentiate scheduled and adhoc flows, so that downstream workloads can also prevent the emission of metrics for these flows that are not tracked.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 1]
```diff
@@ -1014,7 +1014,7 @@ public class ConfigurationKeys {
   public static final String DATASET_BASE_OUTPUT_PATH_KEY = "gobblin.flow.dataset.baseOutputPath";
   public static final String DATASET_COMBINE_KEY = "gobblin.flow.dataset.combine";
   public static final String WHITELISTED_EDGE_IDS = "gobblin.flow.whitelistedEdgeIds";
-
+  public static final String GOBBLIN_FLOW_ISADHOC = "gobblin.flow.isAdhoc";
```
Comment (phet) [inline gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java:1017]: should we debug a separate namespace for these, which are meant to be internal-only, not to be supplied by a user?  e.g. `gobblin.internal.flow.isAdhoc`

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagerUtils.java [hunk 2]
```diff
@@ -307,6 +307,12 @@ static int getDagQueueId(long flowExecutionId, int numThreads) {
     return (int) (flowExecutionId % numThreads);
   }
 
+  static boolean isDagFromAdhocFlow(Dag<JobExecutionPlan> dag) {
+    // Every dag should have at least one node, and every dag will clone the job configurations which include the schedule
+    // defaults to false (so metrics are still tracked) if the dag property is not configured due to old dags
+    return ConfigUtils.getBoolean(dag.getStartNodes().get(0).getValue().getJobSpec().getConfig(), ConfigurationKeys.GOBBLIN_FLOW_ISADHOC,false);
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagerUtils.java:313]: if `DagMgrUtils` doesn't yet have an abstraction for getting a DAG's job config, we might consider one:
```
.getStartNodes().get(0).getValue().getJobSpec().getConfig()
```

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java [hunk 3]
```diff
@@ -237,7 +238,9 @@ public void orchestrate(Spec spec) throws Exception {
       if (!canRun(flowName, flowGroup, allowConcurrentExecution)) {
         _log.warn("Another instance of flowGroup: {}, flowName: {} running; Skipping flow execution since "
             + "concurrent executions are disabled for this flow.", flowGroup, flowName);
-        flowGauges.get(spec.getUri().toString()).setState(CompiledState.SKIPPED);
+        if (flowGauges.containsKey(spec.getUri().toString())) {
+          flowGauges.get(spec.getUri().toString()).setState(CompiledState.SKIPPED);
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java:242]: consider abstracting, e.g. to:
```
conditionallyUpdateFlowGaugeSpecState(flowGauges, spec, state)
```
OR
```
updateSpecStateIfFlowGauge(spec, state, flowGauges)
```

**Review by phet — COMMENTED** [review]
nice tests, BTW!

**Review by phet — APPROVED** [review]

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java [hunk 4]
```diff
@@ -402,6 +402,17 @@ private void deleteFromExecutor(Spec spec, Properties headers) {
     }
   }
 
+  /**
```
[the remark this answers is not in the document]
Comment (ZihanLi58) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java:405]: Synced offline and seems the types of guage  are different and make it difficult to merge two methods, no need to address in this PR

**Review by ZihanLi58 — APPROVED** [review]
+1

**Pull request merged**